crates/runtime/src/instance/allocator/memfd.rs

```diff
@@ -0,0 +1,258 @@
+        // image memfd.
+        unsafe {
+            rustix::io::madvise(
+                self.base as *mut c_void,
+                self.static_size,
+                rustix::io::Advice::LinuxDontNeed,
+            )?;
+        }
+
+        // mprotect the region beyond the initial heap size back to PROT_NONE.
+        self.protect_past_initial_size()?;
+        self.dirty = false;
+        Ok(())
+    }
+
+    fn protect_past_initial_size(&self) -> Result<()> {
+        let mprotect_start = self.base + self.initial_size;
+        let mprotect_len = self.static_size - self.initial_size;
+        if mprotect_len > 0 {
+            unsafe {
+                rustix::io::mprotect(
+                    mprotect_start as *mut _,
+                    mprotect_len,
+                    rustix::io::MprotectFlags::empty(),
+                )?;
+            }
+        }
+
+        Ok(())
+    }
+
+    pub(crate) fn has_image(&self) -> bool {
+        self.image.is_some()
+    }
+
+    pub(crate) fn is_dirty(&self) -> bool {
+        self.dirty
+    }
+}
+
+#[cfg(feature = "memfd-allocator")]
+impl Drop for MemFdSlot {
+    fn drop(&mut self) {
+        // The MemFdSlot may be dropped if there is an error during
+        // instantiation: for example, if a memory-growth limiter
+        // disallows a guest from having a memory of a certain size,
+        // after we've already initialized the MemFdSlot.
+        //
+        // We need to return this region of the large pool mmap to a
+        // safe state (with no module-specific mappings). The
+        // MemFdSlot will not be returned to the MemoryPool, so a new
+        // MemFdSlot will be created and overwrite the mappings anyway
+        // on the slot's next use; but for safety and to avoid
+        // resource leaks it's better not to have stale mappings to a
+        // possibly-otherwise-dead module's image.
+        //
+        // To "wipe the slate clean", let's do a mmap of anonymous
+        // memory over the whole region, with PROT_NONE. Note that we
+        // *can't* simply munmap, because that leaves a hole in the
+        // middle of the pooling allocator's big memory area that some
+        // other random mmap may swoop in and take, to be trampled
+        // over by the next MemFdSlot later.
+        //
+        // Since we're in drop(), we can't sanely return an error if
+        // this mmap fails. Let's ignore the failure if so; the next
+        // MemFdSlot to be created for this slot will try to overwrite
+        // the existing stale mappings, and return a failure properly
+        // if we still cannot map new memory.
+        unsafe {
+            let _ = rustix::io::mmap_anonymous(
+                self.base as *mut _,
+                self.static_size,
+                rustix::io::ProtFlags::empty(),
+                rustix::io::MapFlags::FIXED | rustix::io::MapFlags::NORESERVE,
+            );
+        }
```

Comment:
Ah, so I *think* this bug does not exist, because on `instantiate`, in the different-module case, we remap all memory from the start of heap up to `static_size` (in practice 6GiB) with anonymous zeroes ([here](https://github.com/bytecodealliance/wasmtime/pull/3697/files#diff-1dfa6bf73de63a505980e54c117396b2afb02acda2e4490b750f94f12d12c150R133-R137)) -- this will wipe all previous mappings. We then `mprotect` everything beyond the initial heap size to `PROT_NONE`.

Re: the other bits (sharing clearing logic) that's a good point though and I'll clean this up a bit!

Reply:
... and, actually there was a bug, but a bit more subtle: the initial anonymous mmap was surrounded by `if initial_size > 0` rather than a hypothetical `if static_size > 0`. The former may be false if the memory starts with zero size then grows; the latter is always true in any valid config (so we can make it unconditional). Thanks for poking at this!

Reply:
Fixed!